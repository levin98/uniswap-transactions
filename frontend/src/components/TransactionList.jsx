import React, { useState, useEffect } from 'react';
import { Card, Table, Alert, Pagination, Divider } from 'antd';
import TransactionListForm from './TransactionListForm';
import TransactionStatistics from './TransactionStatistics';

const columns = [
    {
        title: 'Transaction Hash',
        dataIndex: 'txn_hash',
        key: 'txn_hash',
    },
    {
        title: 'Block Number',
        dataIndex: 'block_number',
        key: 'block_number',
    },
    {
        title: 'Timestamp',
        dataIndex: 'txn_timestamp',
        key: 'txn_timestamp',
    },
    {
        title: 'Transaction Fee',
        dataIndex: 'gas_used',
        key: 'gas_used',
        render: (gas_used, row) => {
            return `${(parseFloat(gas_used) * parseFloat(row.gas_price)) * Math.pow(10, -18)} ETH`
        }
    }
]

const TransactionList = () => {
    const [data, setData] = useState([]);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState('');

    const [hash, setHash] = useState('');
    const [from, setFrom] = useState('');
    const [to, setTo] = useState('');

    const [page, setPage] = useState(1);
    const [pageSize, setPageSize] = useState(50);
    const [count, setCount] = useState(0);

    const getTransactions = async (h = hash, f = from, t = to) => {
        setLoading(true);
        setError('');
        try {
            if (h !== "") {
                const response = await fetch(`http://localhost:5000/transaction?hash=${h}`);
                const data = await response.json();
                if (response.status === 404) {
                    setError("Transaction not found");
                    setData([]);
                    setCount(0);
                    return;
                }
                setData([data]);
                setCount(1);
            } else {
                const response = await fetch(`http://localhost:5000/transactions?page=${page}&pageSize=${pageSize}&from=${f}&to=${t}`);
                const data = await response.json();
                setData(data.results);
                setCount(data.pagination.count);
            }
        } catch (e) {
            setError(e.message);
        } finally {
            setLoading(false);
        }
    }

    useEffect(() => {
        getTransactions();
    }, [page, pageSize]);

    const onChangePagination = (page, pageSize) => {
        setPage(page);
        if (pageSize) {
            setPageSize(pageSize);
        }
    }

    const onSubmit = async (values) => {
        setHash(values.hash);
        if (values.timerange) {
            setFrom(values.timerange[0].unix());
            setTo(values.timerange[1].unix());
        }
        await getTransactions(values.hash, values.timerange ? values.timerange[0].unix() : '', values.timerange ? values.timerange[1].unix() : '');
    }

    return (
        <Card className="h-[90vh] w-[90vw]">
            <TransactionStatistics transactionList={data} />
            <Divider />
            <TransactionListForm onSubmit={onSubmit} />
            <br />
            {error && <Alert message={error} type="error" showIcon style={{ marginBottom: '10px' }} />}
            <div className='h-[50vh] overflow-auto'>
                <Table rowKey="txn_hash" columns={columns} loading={loading} dataSource={data} pagination={false} />
            </div>
            <br />
            <div className='text-right'>
                <Pagination current={page} pageSize={pageSize} total={count} onChange={onChangePagination} pageSizeOptions={[10, 20, 50, 100]} showSizeChanger />
            </div>
        </Card>
    );
}

export default TransactionList;