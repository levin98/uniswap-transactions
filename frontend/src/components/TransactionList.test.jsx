import React from 'react';
import { render, screen } from '@testing-library/react';
import TransactionList from './TransactionList';
import "../mock"

// eslint-disable-next-line react/display-name
jest.mock('./TransactionStatistics', () => () => <div data-testid="transaction-statistics">Mocked Transaction Statistics</div>);

// eslint-disable-next-line react/display-name
jest.mock('./TransactionListForm', () => () => <div data-testid="transaction-list-form">Mocked Transaction List Form</div>);

jest.spyOn(window, "fetch").mockImplementation(() => ({}));

test('renders statistics', async () => {
    const { container } = render(<TransactionList />);

    const list = await screen.findByTestId("transaction-list-form");
    const statistics = await screen.findByTestId("transaction-statistics");
    expect(list).toBeInTheDocument();
    expect(statistics).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});
