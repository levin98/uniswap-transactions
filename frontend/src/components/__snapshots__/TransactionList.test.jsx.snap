// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders statistics 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered h-[90vh] w-[90vw] css-dev-only-do-not-override-k7429z"
  >
    <div
      class="ant-card-body"
    >
      <div
        data-testid="transaction-statistics"
      >
        Mocked Transaction Statistics
      </div>
      <div
        class="ant-divider css-dev-only-do-not-override-k7429z ant-divider-horizontal"
        role="separator"
      />
      <div
        data-testid="transaction-list-form"
      >
        Mocked Transaction List Form
      </div>
      <br />
      <div
        class="h-[50vh] overflow-auto"
      >
        <div
          class="ant-table-wrapper css-dev-only-do-not-override-k7429z"
        >
          <div
            class="ant-spin-nested-loading css-dev-only-do-not-override-k7429z"
          >
            <div
              class="ant-spin-container"
            >
              <div
                class="ant-table ant-table-empty"
              >
                <div
                  class="ant-table-container"
                >
                  <div
                    class="ant-table-content"
                  >
                    <table
                      style="table-layout: auto;"
                    >
                      <colgroup />
                      <thead
                        class="ant-table-thead"
                      >
                        <tr>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Transaction Hash
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Block Number
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Timestamp
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Transaction Fee
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="ant-table-tbody"
                      >
                        <tr
                          class="ant-table-placeholder"
                        >
                          <td
                            class="ant-table-cell"
                            colspan="4"
                          >
                            <div
                              class="css-dev-only-do-not-override-k7429z ant-empty ant-empty-normal"
                            >
                              <div
                                class="ant-empty-image"
                              >
                                <svg
                                  height="41"
                                  viewBox="0 0 64 41"
                                  width="64"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <g
                                    fill="none"
                                    fill-rule="evenodd"
                                    transform="translate(0 1)"
                                  >
                                    <ellipse
                                      cx="32"
                                      cy="33"
                                      fill="#f5f5f5"
                                      rx="32"
                                      ry="7"
                                    />
                                    <g
                                      fill-rule="nonzero"
                                      stroke="#d9d9d9"
                                    >
                                      <path
                                        d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                      />
                                      <path
                                        d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                        fill="#fafafa"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </div>
                              <div
                                class="ant-empty-description"
                              >
                                No data
                              </div>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div
        class="text-right"
      >
        <ul
          class="ant-pagination css-dev-only-do-not-override-k7429z"
        >
          <li
            aria-disabled="true"
            class="ant-pagination-prev ant-pagination-disabled"
            title="Previous Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="left"
                class="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-disabled"
            tabindex="0"
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            aria-disabled="true"
            class="ant-pagination-next ant-pagination-disabled"
            title="Next Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="right"
                class="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-options"
          >
            <div
              aria-label="Page Size"
              class="ant-select ant-pagination-options-size-changer css-dev-only-do-not-override-k7429z ant-select-single ant-select-show-arrow"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-label="Page Size"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="50 / page"
                >
                  50 / page
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
